### PR TITLE
Improve extractor performance by buffering filtered transactions

### PR DESCRIPTION
**Pull request by devgrok**
The mapAsync in Extractor.scala doesn't perform the action of the Writer in parallel with the previous steps in the stream:
   processing grpc -> transform -> filter -> mapAsync .....> updateOffset 

akka's `mapAsync` with `parallelism=1` creates a buffer of size=1 which holds the element being processed, so when waiting for the callback the buffer is full which causes backpressure and triggers the stream processing to stop. 
This effectively means that the retrieving the transaction + transformation and filtering happens between calls to the Writer.
By adding a small buffer, it allows this processing to occur in the background as needed.

@ASX we've seen up to 100ms per 1s being wasted by the Writer waiting for input, this change effectively removes that wait in the load testing we've performed.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [x] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -76,8 +76,19 @@ We have a single script to build most targets and run the tests. On Linux and Ma
...
+#### Windows
+On Windows you can just run `.\build.ps1` to do a complete build that downloads all the required tools and cleans up any potentially broken builds and has more of a chance of working than invoking bazel directly as suggested above.
+
+Note: you'll need to add the full path to `<proj>\dev-env\windows\bin` to your User or System environment varaibles (a SET PATH=%PATH%;.... won't work as the dadew script reloads then environment varibles from the registry.
+
+Note 2: the build.ps1 script does a one off init'ing `.bazelrc.local` to the following, that may be required for windows builds to work properly:
+```
+build --config windows
+```
+
```
Comment (stefanobaghino-da) [inline README.md:91]: 🙇🏻

File: extractor/src/main/scala/com/digitalasset/extractor/Extractor.scala [hunk 2]
```diff
@@ -188,6 +188,11 @@ class Extractor[T](config: ExtractorConfig, target: T)(
           .collect {
             Function.unlift(convertTransactionTree(parties, requestedTemplateIds))
           }
+          // mapAsync with par=1 creates a buffer size=1 which holds the element being processed, when waiting on callback
+          // the buffer is full and triggers backpressure. 
+          // A buffer after the transforming & filtering is required to allow it to execute in parralel and not be blocked waiting 
+          // for the async call to return.
```
Comment (stefanobaghino-da) [inline extractor/src/main/scala/com/digitalasset/extractor/Extractor.scala:194]: Thanks for the descriptive comment!

File: extractor/src/main/scala/com/digitalasset/extractor/Extractor.scala [hunk 3]
```diff
@@ -47,7 +47,7 @@ class Extractor[T](config: ExtractorConfig, target: T)(
   private val tokenHolder = config.accessTokenFile.map(new TokenHolder(_))
   private val parties: Set[String] = config.parties.toSet.map(identity)
 
-  implicit val system: ActorSystem = ActorSystem()
+  lazy implicit val system: ActorSystem = ActorSystem()
```
Comment (stefanobaghino-da) [inline extractor/src/main/scala/com/digitalasset/extractor/Extractor.scala:50]: > Make ActorSystem lazy so subclasses can override to customize it.

I'm not sure I understand how laziness affects the possibility of overriding this, can you explain, please?

**Review by stefanobaghino-da — COMMENTED** [review]
Thanks for the contribution. I'm not 100% convinced by the `lazy` modifier to allow sub-classes to override the `ActorSystem`.

**Review by S11001001 — APPROVED** [review]
Looks good, thanks.

**Comment by S11001001** [comment]
* [ ] forward-port to main

File: extractor/src/main/scala/com/digitalasset/extractor/Extractor.scala (hunk 3 above)
Comment (S11001001) [reply extractor/src/main/scala/com/digitalasset/extractor/Extractor.scala:50]: It affects whether references to `system` can be overridden _in time_.

```scala
scala> class Foo {val x = ""; val y = x}; class Bar extends Foo {override val x = "please be y"}; (new Bar).y
class Foo
class Bar
val res1: String = null

scala> class Foo {lazy val x = ""; val y = x}; class Bar extends Foo {override lazy val x = "please be y"}; (new Bar).y
class Foo
class Bar
val res3: String = please be y
```

File: README.md (hunk 1 above)
Comment (devgrok) [reply README.md:91]: Happy to remove - by having it lazy stops an orphaned `ActorSystem` being created when it's overridden.
Was your concern around overriding it or the `lazy`?

File: extractor/src/main/scala/com/digitalasset/extractor/Extractor.scala (hunk 3 above)
Comment (devgrok) [reply extractor/src/main/scala/com/digitalasset/extractor/Extractor.scala:50]: I had found an interesting way to override it, but the base class's value is still initialized before being overridden and that  value is not referenceable.
So in this case, an extra `ActorSystem` is created and dosn't shutdown.
```scala
class Foo {val x = ""; val y = x}; class Bar(override val x: String = "please be y") extends Foo {}; (new Bar).y
```

**Pull request merged**